components/br-stream/src/endpoint.rs

```diff
@@ -134,17 +135,16 @@ where
             }
         }
     }
-    // TODO use a more efficent encode kv event format
     // TODO move this function to a indepentent module.
-    pub fn encode_event(key: &[u8], value: &[u8]) -> Vec<u8> {
-        let mut buf = vec![];
-        let key_len = (key.len() as u32).to_ne_bytes();
-        let val_len = value.len().to_ne_bytes();
-        buf.extend_from_slice(&key_len);
-        buf.extend_from_slice(key);
-        buf.extend_from_slice(&val_len);
-        buf.extend_from_slice(value);
-        buf
+    pub fn encode_event<'e>(key: &'e [u8], value: &'e [u8]) -> [impl AsRef<[u8]> + 'e; 4] {
```

Comment:
(BTW, I would prefer make it something like `encode_event_to(w: impl Write, key: &[u8], value: &[u8])`, the current version is a little... uh... magical...)

Reply:
@YuJuncen the problem is, the `w` isn't a simple `impl Write` but an `AsyncWrite` plus a SHA-256 sink. (why are we using async to write to *local* temp file 😂)

Reply:
> the w isn't a simple impl Write but an AsyncWrite plus a SHA-256 sink

I guess we can make something like `Sha256Io<T>`, which likes `Sha256Reader` but impls `Write` / `Read` / `AsyncRead` / `AsyncWrite` when the inner type implements corresponding traits. 🤔 (Probably not for this PR)

> why are we using async to write to local temp file 😂

Just PTSD of Sync I/O in nodejs. 😂